Sources/LanguageServerProtocol/AsyncQueue.swift

```diff
@@ -0,0 +1,74 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+/// Abstraction layer so we can store a heterogeneous collection of tasks in an
+/// array.
+private protocol AnyTask: Sendable {
+  func waitForCompletion() async
+}
+
+extension Task: AnyTask {
+  func waitForCompletion() async {
+    _ = try? await value
+  }
+}
+
+/// A serial queue that allows the execution of asyncronous blocks of code.
+public final class AsyncQueue {
+  /// Lock guarding `pendingTasks`.
+  private let pendingTasksLock = NSLock()
+
+  /// Tasks that haven't finished yet.
+  private var pendingTasks = [UUID: AnyTask]()
+
+  public init() {
+    self.pendingTasksLock.name = "AsyncQueue"
+  }
+
+  /// Schedule a new closure to be executed on the queue.
+  ///
+  /// All previously added tasks are guaranteed to finished executing before
+  /// this closure gets executed.
+  @discardableResult
+  public func async<Success: Sendable>(
+    priority: TaskPriority? = nil,
+    @_inheritActorContext operation: @escaping @Sendable () async -> Success
+  ) -> Task<Success, Never> {
+    let id = UUID()
+
+    return pendingTasksLock.withLock {
+      let tasksToAwait = pendingTasks.values
+
+      let task = Task<Success, Never>(priority: priority) { [tasksToAwait] in
+        for task in tasksToAwait {
+          await task.waitForCompletion()
+        }
```

Comment:
It's entirely possible I'm missing something here, but wouldn't it be sufficient to store only the last task, and await on that? Since each task awaits on the previous one, wouldn't that provide the same ordering guarantee?

Reply:
Hmm, yes, that should be sufficient. Good catch.